tests/integration/lifecycle/test_craftctl.py

```diff
@@ -0,0 +1,428 @@
+    )
+    actions = lf.plan(Step.PRIME)
+    assert actions == [
+        Action("foo", Step.PULL),
+        Action("foo", Step.OVERLAY),
+        Action("foo", Step.BUILD),
+        Action("foo", Step.STAGE),
+        Action("foo", Step.PRIME),
+    ]
+    with lf.action_executor() as ctx:
+        ctx.execute(actions[0])
+        captured = capfd.readouterr()
+        assert captured.out == "pull step\n"
+        assert Path("parts/foo/src/foo.txt").exists()
+        assert Path("parts/foo/install/foo.txt").exists() is False
+        assert Path("stage/foo.txt").exists() is False
+        assert Path("prime/foo.txt").exists() is False
+
+        ctx.execute(actions[1])
+        captured = capfd.readouterr()
+        Path("parts/foo/layer/ovl.txt").touch()
+        assert captured.out == "overlay step\n"
+
+        ctx.execute(actions[2])
+        captured = capfd.readouterr()
+        assert captured.out == "build step\n"
+        assert Path("parts/foo/install/foo.txt").exists()
+        assert Path("stage/foo.txt").exists() is False
+        assert Path("stage/ovl.txt").exists() is False
+        assert Path("prime/foo.txt").exists() is False
+        assert Path("prime/ovl.txt").exists() is False
+
+        ctx.execute(actions[3])
+        captured = capfd.readouterr()
+        assert captured.out == "stage step\n"
+        assert Path("stage/foo.txt").exists()
+        assert Path("stage/ovl.txt").exists()
+        assert Path("prime/foo.txt").exists() is False
+        assert Path("prime/ovl.txt").exists() is False
+
+        ctx.execute(actions[4])
+        captured = capfd.readouterr()
+        assert captured.out == "prime step\n"
+        assert Path("prime/foo.txt").exists()
+        assert Path("prime/ovl.txt").exists()
+
+
+def test_craftctl_default_argments(new_dir):
+    parts_yaml = textwrap.dedent(
+        """\
+        parts:
+          foo:
+            plugin: nil
+            override-pull: craftctl default argument
+        """
+    )
+    parts = yaml.safe_load(parts_yaml)
+
+    lf = craft_parts.LifecycleManager(
+        parts,
+        application_name="test_ctl",
+        cache_dir=new_dir,
+        base_layer_dir=new_dir,
+        base_layer_hash=b"hash",
+    )
+    with pytest.raises(errors.InvalidControlAPICall) as raised:
+        with lf.action_executor() as ctx:
+            ctx.execute(Action("foo", Step.PULL))
+
+    assert raised.value.message == "invalid arguments to command 'default'"
+
+
+def test_craftctl_set(new_dir):
+    parts_yaml = textwrap.dedent(
+        """\
+        parts:
+          foo:
+            plugin: nil
+            override-pull: |
+              craftctl set myvar=myvalue
+        """
+    )
+    parts = yaml.safe_load(parts_yaml)
+
+    lf = craft_parts.LifecycleManager(
+        parts,
+        application_name="test_set",
+        cache_dir=new_dir,
+        project_vars={"myvar": ""},
+    )
+    with lf.action_executor() as ctx:
+        ctx.execute(Action("foo", Step.PULL))
```

Comment:
we need to test later running Step.BUILD and retrieving the value of a previously stored Step.PULL (which would not run due to the step caching)

Reply:
Ah I see the problem here. We need to store project vars in step state. Will implement and test.

Reply:
Store project variables in step state along with update flag. On a new execution the sequencer reads the state and passes any project var states in skip actions. Executor updates the variables if action is skipped.